File: ios/Demo/MekuriDemo/DemoSpeechBubble.swift
import SwiftUI

/// Where a bubble's tail points, as a fraction of the bubble's own width.
enum DemoBubbleTail {
    case bottomLeading
    case bottomTrailing
    case topLeading
    case topTrailing
    case none
}

/// An ellipse with a tail. The tail is unioned into the ellipse so the
/// stroke runs around the outside only.
struct DemoBubbleShape: Shape {
    let tail: DemoBubbleTail

    func path(in rect: CGRect) -> Path {
        let inset: CGFloat = 12
        let body = rect.insetBy(dx: 0, dy: inset)
        var path = Path(ellipseIn: body)
        switch self.tail {
        case .none:
            return path
        case .bottomLeading:
            path.addPath(self.tailPath(from: CGPoint(x: body.minX + body.width * 0.30, y: body.maxY - 6), tip: CGPoint(x: body.minX + body.width * 0.12, y: rect.maxY), toward: CGPoint(x: body.minX + body.width * 0.45, y: body.maxY - 4)))
        case .bottomTrailing:
            path.addPath(self.tailPath(from: CGPoint(x: body.maxX - body.width * 0.30, y: body.maxY - 6), tip: CGPoint(x: body.maxX - body.width * 0.12, y: rect.maxY), toward: CGPoint(x: body.maxX - body.width * 0.45, y: body.maxY - 4)))
        case .topLeading:
            path.addPath(self.tailPath(from: CGPoint(x: body.minX + body.width * 0.30, y: body.minY + 6), tip: CGPoint(x: body.minX + body.width * 0.12, y: rect.minY), toward: CGPoint(x: body.minX + body.width * 0.45, y: body.minY + 4)))
        case .topTrailing:
            path.addPath(self.tailPath(from: CGPoint(x: body.maxX - body.width * 0.30, y: body.minY + 6), tip: CGPoint(x: body.maxX - body.width * 0.12, y: rect.minY), toward: CGPoint(x: body.maxX - body.width * 0.45, y: body.minY + 4)))
        }
        return path
    }

    private func tailPath(from: CGPoint, tip: CGPoint, toward: CGPoint) -> Path {
        var path = Path()
        path.move(to: from)
        path.addLine(to: tip)
        path.addLine(to: toward)
        path.closeSubpath()
        return path
    }
}

/// Speech text in a bubble. `size` scales the type with the page.
struct DemoSpeechBubble: View {
    let text: String
    var tail: DemoBubbleTail = .bottomLeading
    var size: CGFloat = 15

    var body: some View {
        Text(self.text)
            .font(DemoInk.speech(self.size))
            .multilineTextAlignment(.center)
            .foregroundStyle(DemoInk.ink)
            .padding(.horizontal, self.size * 1.4)
            .padding(.vertical, self.size * 1.5)
            .background(
                DemoBubbleShape(tail: self.tail)
                    .fill(DemoInk.paper)
                    .stroke(DemoInk.ink, lineWidth: DemoInk.bubbleStroke)
            )
    }
}

/// A jagged burst for a sound effect.
struct DemoBurstShape: Shape {
    var points = 14
    var depth: CGFloat = 0.22

    func path(in rect: CGRect) -> Path {
        let center = CGPoint(x: rect.midX, y: rect.midY)
        let outer = CGSize(width: rect.width / 2, height: rect.height / 2)
        var path = Path()
        for index in 0..<(self.points * 2) {
            let angle = CGFloat(index) / CGFloat(self.points * 2) * .pi * 2 - .pi / 2
            let scale: CGFloat = index.isMultiple(of: 2) ? 1 : 1 - self.depth
            let point = CGPoint(x: center.x + cos(angle) * outer.width * scale, y: center.y + sin(angle) * outer.height * scale)
            if index == 0 { path.move(to: point) } else { path.addLine(to: point) }
        }
        path.closeSubpath()
        return path
    }
}

/// Sound-effect lettering, optionally inside a burst.
struct DemoSoundEffect: View {
    let text: String
    var size: CGFloat = 40
    var burst = false

    var body: some View {
        Text(self.text)
            .font(DemoInk.sfx(self.size))
            .foregroundStyle(DemoInk.ink)
            .padding(.horizontal, self.burst ? self.size * 0.9 : 0)
            .padding(.vertical, self.burst ? self.size * 0.7 : 0)
            .background {
                if self.burst {
                    DemoBurstShape()
                        .fill(DemoInk.paper)
                        .stroke(DemoInk.ink, lineWidth: DemoInk.panelStroke)
                }
            }
    }
}

File: android/demo/src/main/kotlin/studio/anicanon/mekuri/demo/DemoSpeechBubble.kt
package studio.anicanon.mekuri.demo

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.text.BasicText
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import kotlin.math.cos
import kotlin.math.sin

/** Where a bubble's tail points, as a fraction of the bubble's own width. */
enum class DemoBubbleTail {
    BottomLeading,
    BottomTrailing,
    TopLeading,
    None,
}

/** Speech text in a bubble. [size] scales the type with the page. */
@Composable
fun DemoSpeechBubble(
    text: String,
    modifier: Modifier = Modifier,
    tail: DemoBubbleTail = DemoBubbleTail.BottomLeading,
    size: Float = 15f,
) {
    Box(
        modifier
            .drawBehind {
                val path = bubblePath(Rect(Offset.Zero, this.size), tail)
                drawPath(path, DemoInk.Paper)
                drawPath(path, DemoInk.Ink, style = Stroke(width = 2.5f))
            }
            .padding(horizontal = (size * 1.4f).dp, vertical = (size * 1.5f).dp),
    ) {
        BasicText(text = text, style = DemoInk.speech(size).copy(textAlign = TextAlign.Center))
    }
}

/** Sound-effect lettering, optionally inside a burst. */
@Composable
fun DemoSoundEffect(
    text: String,
    modifier: Modifier = Modifier,
    size: Float = 40f,
    burst: Boolean = false,
) {
    val padded = if (burst) {
        modifier
            .drawBehind {
                val path = burstPath(Rect(Offset.Zero, this.size))
                drawPath(path, DemoInk.Paper)
                drawPath(path, DemoInk.Ink, style = Stroke(width = 3f))
            }
            .padding(horizontal = (size * 0.9f).dp, vertical = (size * 0.7f).dp)
    } else {
        modifier
    }
    BasicText(text = text, modifier = padded, style = DemoInk.sfx(size))
}

/**
 * An ellipse with a tail. The tail is part of the same path; the stroke runs
 * around the outside only.
 */
private fun bubblePath(rect: Rect, tail: DemoBubbleTail): Path {
    val inset = 12f
    val body = Rect(rect.left, rect.top + inset, rect.right, rect.bottom - inset)
    val path = Path().apply { addOval(body) }
    when (tail) {
        DemoBubbleTail.None -> Unit
        DemoBubbleTail.BottomLeading -> path.addTail(
            from = Offset(body.left + body.width * 0.30f, body.bottom - 6f),
            tip = Offset(body.left + body.width * 0.12f, rect.bottom),
            toward = Offset(body.left + body.width * 0.45f, body.bottom - 4f),
        )
        DemoBubbleTail.BottomTrailing -> path.addTail(
            from = Offset(body.right - body.width * 0.30f, body.bottom - 6f),
            tip = Offset(body.right - body.width * 0.12f, rect.bottom),
            toward = Offset(body.right - body.width * 0.45f, body.bottom - 4f),
        )
        DemoBubbleTail.TopLeading -> path.addTail(
            from = Offset(body.left + body.width * 0.30f, body.top + 6f),
            tip = Offset(body.left + body.width * 0.12f, rect.top),
            toward = Offset(body.left + body.width * 0.45f, body.top + 4f),
        )
    }
    return path
}

private fun Path.addTail(from: Offset, tip: Offset, toward: Offset) {
    moveTo(from.x, from.y)
    lineTo(tip.x, tip.y)
    lineTo(toward.x, toward.y)
    close()
}

/** A jagged burst for a sound effect. */
private fun burstPath(rect: Rect, points: Int = 14, depth: Float = 0.22f): Path {
    val path = Path()
    val centerX = rect.center.x
    val centerY = rect.center.y
    val outerX = rect.width / 2
    val outerY = rect.height / 2
    for (index in 0 until points * 2) {
        val angle = index.toFloat() / (points * 2).toFloat() * Math.PI.toFloat() * 2f - Math.PI.toFloat() / 2f
        val scale = if (index % 2 == 0) 1f else 1f - depth
        val x = centerX + cos(angle) * outerX * scale
        val y = centerY + sin(angle) * outerY * scale
        if (index == 0) path.moveTo(x, y) else path.lineTo(x, y)
    }
    path.close()
    return path
}
